lib/checkleakautovar.cpp

```diff
@@ -141,7 +141,7 @@ void VarInfo::possibleUsageAll(const std::string &functionName)
 }
 
 
-void CheckLeakAutoVar::leakError(const Token *tok, const std::string &varname, int type)
+void CheckLeakAutoVar::leakError(const Token *tok, const std::string &varname, int type) const
```

Comment:
This does not feel natural to me. The `mErrorLogger->reportError` is called and that has side effects. it seems the checkmemleak methods should not be const.

Reply:
I don't have a good feeling about this check. If the intention is not to follow good programming practice or make the code more robust..

Should it be moved to an addon perhaps?

Reply:
I am fine with this - the check is also `inconclusive` so that's not something you should apply in production code unless properly reviewed.

In case of `ErrorLogger` propagating (or rather not) the `const` properly will make several things mutable which shouldn't. So we possibly need to rework the way we report errors. An idea is collecting them in a mutable context and only send them to the logger at the end. As the suppressions also need to be mutable and they are also tied to the logger that make sense to work those together. I haven't looked into it yet.

Also if the code is actually sound and side effect-free and there's an interface that claims to be `const` we will suggest this even if the implementation behind that interface isn't. But you have the same problem with any annotation which might not actually be "logically" correct.

That's why we need more checks to get rid of pointers and non-`const` stuff as well as detecting code that violates `constness`. Also using the proposed `safe_ptr` will expose this in out code base.

Reply:
I have changed my mind.

I think it's ok to keep the checker. And making these methods that report errors const seems natural. I can agree that it does not effect the "natural state" of the checker nor the ErrorLogger.

Reply:
It's complex. On one side there's enforcing the annotations the compiler/standard offers you. But on the other side there's your actual definition what `const` on an object should be.

I am wondering if we should turn any existing bailouts in this check into their own warnings since that would be violations of the "logical constness".

Reply:
it depends if the bailouts usually trigger when the natual state is changed then personally I would just keep the bailout even though that can cause false negatives.

Feel free to remove a bailout in a separate PR and check what the results will be in daca.

Reply:
My suggestion was to issue an separate warning on that bailout and not remove them. We could start with making those debug warnings similar to the valueflow bailouts so they will be collected in daca.

Reply:
yes I think that sounds interesting